exonum/benches/criterion/storage.rs

```diff
@@ -45,58 +52,119 @@ fn generate_random_kv(len: usize) -> Vec<([u8; KEY_SIZE], Vec<u8>)> {
             rng.fill_bytes(&mut k);
         }
         exists_keys.insert(k);
-        (k, v)
+        (Hash::new(k), v)
     };
 
     (0..len).map(kv_generator).collect::<Vec<_>>()
 }
 
-fn merkle_table_insertion<T: Database>(b: &mut Bencher, db: &T) {
+fn proof_list_append<T: Database>(b: &mut Bencher, db: &T, len: &usize) {
     let mut rng = XorShiftRng::from_seed([200; 16]);
     let mut storage = db.fork();
     let mut table = ProofListIndex::new(NAME, &mut storage);
 
-    b.iter(|| {
-        let v_generator = |_| {
-            let mut chunk = vec![0; 10];
-            rng.fill_bytes(&mut chunk);
-            chunk
-        };
-
-        for item in (0..1000).map(v_generator) {
-            table.push(item);
-        }
-    });
+    b.iter_with_setup(
+        || {
+            (0..*len)
+                .map(|_| {
+                    let mut chunk = vec![0; 64];
+                    rng.fill_bytes(&mut chunk);
+                    chunk
+                }).collect::<Vec<_>>()
+        },
+        |data| {
+            for item in data {
+                table.push(item);
+            }
+        },
+    );
+}
+
+fn proof_map_insert_without_merge<T: Database>(b: &mut Bencher, db: &T, len: &usize) {
+    let data = generate_random_kv(*len);
+    b.iter_with_setup(
+        || db.fork(),
+        |mut storage| {
+            let mut table = ProofMapIndex::new(NAME, &mut storage);
+            for item in &data {
+                table.put(&item.0, item.1.clone());
+            }
+        },
+    );
+}
+
+fn proof_map_insert_with_merge<T: Database>(b: &mut Bencher, db: &T, len: &usize) {
+    let data = generate_random_kv(*len);
+    b.iter_with_large_setup(
+        || {
+            let patch = {
+                let mut fork = db.fork();
+                {
+                    let mut table: ProofMapIndex<_, Hash, Vec<u8>> =
+                        ProofMapIndex::new(NAME, &mut fork);
+                    table.clear();
```

Comment:
Why don't we clear the table on each iteration? Is it expensive enough to distort the benchmark?

Reply:
Update: On the second look, I don't really understand what `iter_with_large_setup` tries to accomplish. Have you looked into its docs or implementation? The setup step is supposed to generate some data which is then funneled into iterations. Here, we don't actually generate any data, but rather effectively do a no-op a bunch of times, and then fill the table with the same values (i.e., do something that could be theoretically optimized into no-op on every iteration barring the first one).

It could make sense to generate random values for each iteration and then write them under the same keys; in this case, we reasonably don't need table clearing (although, for surety, we can initially fill the keys before any iterations). *Or* we can generate both keys and values and clear the table on each iteration.

Reply:
Yes `clear` just iterate over the whole keys.